## Table of Contents


1) Import Libraries and Data

2) Data Wrangling for Customers data set

3) Data Quality and Consistency checks for Customers data set

4) Merge custoemr data with orders/products data

5) Export dataframe

### 1) 
Import Libraries and Data

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#define path
path = r'C:\Users\Owner\Documents\Career Foundry\Tasks\Data Immersion Tasks\Instacart Project\2 Data'

Import analysis libraries, as well as your new customer data set as a dataframe.

In [3]:
#import merged product and orders data as df_ord_prod with price ranges, busy counts, and user flags
df_ord_prod = pd.read_pickle(os.path.join(path, 'prepared data', 'order product merge with price ranges-busy counts-user flags.pkl' ))

In [3]:
#import new customer data as df_cust
df_cust = pd.read_csv(os.path.join(path, 'original data', 'customers.csv'))

### 2)
Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [4]:
#check out shape of data
df_cust.shape

(206209, 10)

#### Renaming Columns

In [5]:
#get a look at data frame head for any obvious visual problems
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [6]:
#grab column names to rename columns
df_cust.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [7]:
#rename First Name as 'first_name'
#rename Surnam as 'last_name'
#rename n_dependants as 'num_of_dependants'
df_cust.rename(columns = {'First Name':'first_name'},inplace = True)
df_cust.rename(columns = {'Surnam':'last_name'},inplace = True)
df_cust.rename(columns = {'n_dependants':'num_of_dependants'},inplace = True)

In [8]:
#regrab column names to confirm renaming
df_cust.columns

Index(['user_id', 'first_name', 'last_name', 'Gender', 'STATE', 'Age',
       'date_joined', 'num_of_dependants', 'fam_status', 'income'],
      dtype='object')

### 3)
Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

#### Look for missing values in columns

In [9]:
#search the entire dataframe for Nan values
df_cust.isnull().sum().sum()

11259

In [14]:
#search by columns for Nan values
df_cust['user_id'].isnull().sum()

0

In [15]:
df_cust['first_name'].isnull().sum()

11259

In [16]:
df_cust['last_name'].isnull().sum()

0

In [17]:
df_cust['Gender'].isnull().sum()

0

In [18]:
df_cust['STATE'].isnull().sum()

0

In [19]:
df_cust['Age'].isnull().sum()

0

In [20]:
df_cust['date_joined'].isnull().sum()

0

In [21]:
df_cust['num_of_dependants'].isnull().sum()

0

In [22]:
df_cust['fam_status'].isnull().sum()

0

In [23]:
df_cust['income'].isnull().sum()

0

Confirmed the only Nan values all exist in the 'first_name' column. I will choose to move forward with no action.

#### Check for unusual values in categorical columns (exculding names)

In [10]:
df_cust['Gender'].value_counts(dropna=False)

Male      104067
Female    102142
Name: Gender, dtype: int64

In [11]:
df_cust['STATE'].value_counts(dropna=False)

Delaware                4044
Alabama                 4044
Alaska                  4044
Florida                 4044
Hawaii                  4044
Arkansas                4044
District of Columbia    4044
Colorado                4044
Georgia                 4044
Iowa                    4044
Arizona                 4044
Connecticut             4044
Idaho                   4044
Indiana                 4044
California              4044
Illinois                4044
Wyoming                 4043
North Dakota            4043
Vermont                 4043
Kansas                  4043
Nebraska                4043
Missouri                4043
Texas                   4043
Pennsylvania            4043
Minnesota               4043
Nevada                  4043
Rhode Island            4043
Utah                    4043
New Hampshire           4043
Oklahoma                4043
New York                4043
Oregon                  4043
Tennessee               4043
Washington              4043
Montana       

In [12]:
df_cust['fam_status'].value_counts(dropna=False)

married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: fam_status, dtype: int64

Category column values and counts all look to be reasonable

### Check for any unusual values in numerical data

In [14]:
df_cust.describe()

user_id            Age  num_of_dependants         income
count  206209.000000  206209.000000      206209.000000  206209.000000
mean   103105.000000      49.501646           1.499823   94632.852548
std     59527.555167      18.480962           1.118433   42473.786988
min         1.000000      18.000000           0.000000   25903.000000
25%     51553.000000      33.000000           0.000000   59874.000000
50%    103105.000000      49.000000           1.000000   93547.000000
75%    154657.000000      66.000000           3.000000  124244.000000
max    206209.000000      81.000000           3.000000  593901.000000

#### Check for mixed data types in columns

In [27]:
#get overall info on data types for columns
df_cust.dtypes

user_id               int64
first_name           object
last_name            object
Gender               object
STATE                object
Age                   int64
date_joined          object
num_of_dependants     int64
fam_status           object
income                int64
dtype: object

In [28]:
#search data frame for mixed data types using for loop form 4.5
for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

first_name


Only mixed data types is first_name column. This makes sense as the NaN values would not be considered objects. I will choose to keep these as NaN values.

#### Checking for duplicates

In [29]:
#search for full duplciates in df_cust and store them in df_cust_dupes
df_cust_dupes = df_cust[df_cust.duplicated()]

In [30]:
#check df_cust_dupes for any values
df_cust_dupes.shape

(0, 10)

### 4)
Combine your customer data with the rest of your prepared Instacart data. Tip: Make sure the key columns are of the same data type!

In [31]:
#review headers of each data set df_cust and df_ord_prod
df_cust.head()

user_id first_name last_name  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   num_of_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374

In [4]:
df_ord_prod.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329.0      1.0           1.0                 2.0                8.0   
1  2398795.0      1.0           2.0                 3.0                7.0   
2   473747.0      1.0           3.0                 3.0               12.0   
3  2254736.0      1.0           4.0                 4.0                7.0   
4   431534.0      1.0           5.0                 4.0               15.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                1.0        0.0   
1                    15.0         196                1.0        1.0   
2                    21.0         196                1.0        1.0   
3                    29.0         196                1.0        1.0   
4                    28.0         196                1.0        1.0   

  product_name  ...        price_range     busiest_day  busiest_2_days  \
0         Soda  ...  Mid-range product  Regularly busy  Regularly busy   
1         Soda  ...  Mid-range product  Regularly busy      Least busy   
2         Soda  ...  Mid-range product  Regularly busy      Least busy   
3         Soda  ...  Mid-range product      Least busy      Least busy   
4         Soda  ...  Mid-range product      Least busy      Least busy   

  busiest_period_of_day max_order  loyalty_flag avg_user_price   spend_flag  \
0        Average orders      10.0  New Customer       6.367797  Low Spender   
1        Average orders      10.0  New Customer       6.367797  Low Spender   
2           Most orders      10.0  New Customer       6.367797  Low Spender   
3        Average orders      10.0  New Customer       6.367797  Low Spender   
4           Most orders      10.0  New Customer       6.367797  Low Spender   

   median_days_since_order         frequency_flag  
0                     20.5  Non-frequent customer  
1                     20.5  Non-frequent customer  
2                     20.5  Non-frequent customer  
3                     20.5  Non-frequent customer  
4                     20.5  Non-frequent customer  

[5 rows x 24 columns]

In [5]:
#greab column names for df_ord_prod
df_ord_prod.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', '_merge', 'price_range', 'busiest_day',
       'busiest_2_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'avg_user_price', 'spend_flag', 'median_days_since_order',
       'frequency_flag'],
      dtype='object')

In [6]:
#remove '_merge' column from df_ord_prod so it can be added again in merge
df_ord_prod = df_ord_prod.drop(columns = ['_merge'])

In [32]:
#use merge to join dataframes on shared user_id column
df_cust_prod_ord = df_ord_prod.merge(df_cust, on= ['user_id'], how = 'outer', indicator = True)

In [35]:
df_ord_prod.shape

(32435070, 23)

In [33]:
#check the indicator column in df_cust_prod_ord counts
df_cust_prod_ord['_merge'].value_counts()

both          32435059
left_only           11
right_only           0
Name: _merge, dtype: int64

In [34]:
#take a look at the values from the orders and products data that did not share a user_id with the custoemr data
df_cust_prod_ord[df_cust_prod_ord['_merge']=='left_only']

order_id  user_id  order_number  orders_day_of_week  \
32435059       NaN      NaN           NaN                 NaN   
32435060       NaN      NaN           NaN                 NaN   
32435061       NaN      NaN           NaN                 NaN   
32435062       NaN      NaN           NaN                 NaN   
32435063       NaN      NaN           NaN                 NaN   
32435064       NaN      NaN           NaN                 NaN   
32435065       NaN      NaN           NaN                 NaN   
32435066       NaN      NaN           NaN                 NaN   
32435067       NaN      NaN           NaN                 NaN   
32435068       NaN      NaN           NaN                 NaN   
32435069       NaN      NaN           NaN                 NaN   

          order_hour_of_day  days_since_prior_order  product_id  \
32435059                NaN                     NaN        3630   
32435060                NaN                     NaN        3718   
32435061                NaN                     NaN        7045   
32435062                NaN                     NaN       25383   
32435063                NaN                     NaN       27499   
32435064                NaN                     NaN       36233   
32435065                NaN                     NaN       37703   
32435066                NaN                     NaN       43725   
32435067                NaN                     NaN       45971   
32435068                NaN                     NaN       46625   
32435069                NaN                     NaN       49540   

          add_to_cart_order  reordered  \
32435059                NaN        NaN   
32435060                NaN        NaN   
32435061                NaN        NaN   
32435062                NaN        NaN   
32435063                NaN        NaN   
32435064                NaN        NaN   
32435065                NaN        NaN   
32435066                NaN        NaN   
32435067                NaN        NaN   
32435068                NaN        NaN   
32435069                NaN        NaN   

                                               product_name  ...  first_name  \
32435059                        Protein Granola Apple Crisp  ...         NaN   
32435060                   Wasabi Cheddar Spreadable Cheese  ...         NaN   
32435061             Unpeeled Apricot Halves in Heavy Syrup  ...         NaN   
32435062                                 Chocolate Go Bites  ...         NaN   
32435063                  Non-Dairy Coconut Seven Layer Bar  ...         NaN   
32435064                            Water With Electrolytes  ...         NaN   
32435065  Ultra Sun Blossom Liquid 90 loads Fabric Enhan...  ...         NaN   
32435066                               Sweetart Jelly Beans  ...         NaN   
32435067                         12 Inch Taper Candle White  ...         NaN   
32435068    Single Barrel Kentucky Straight Bourbon Whiskey  ...         NaN   
32435069                            Pure Squeezed  Lemonade  ...         NaN   

          last_name  Gender STATE Age date_joined num_of_dependants  \
32435059        NaN     NaN   NaN NaN         NaN               NaN   
32435060        NaN     NaN   NaN NaN         NaN               NaN   
32435061        NaN     NaN   NaN NaN         NaN               NaN   
32435062        NaN     NaN   NaN NaN         NaN               NaN   
32435063        NaN     NaN   NaN NaN         NaN               NaN   
32435064        NaN     NaN   NaN NaN         NaN               NaN   
32435065        NaN     NaN   NaN NaN         NaN               NaN   
32435066        NaN     NaN   NaN NaN         NaN               NaN   
32435067        NaN     NaN   NaN NaN         NaN               NaN   
32435068        NaN     NaN   NaN NaN         NaN               NaN   
32435069        NaN     NaN   NaN NaN         NaN               NaN   

          fam_status income     _merge  
32435059         NaN    NaN  left_only  
32435060         NaN   

Total rows for customer data was 206209 and orders/products having 32435070. This matches the total of both and left only rows in new dataframe.

The left only rows were the 11 rows of product data that did not share a product id with the orders data. As these products were not found in any orders they will be removed from the data

In [36]:
#redefine df_cust_prod_ord as all values except the 11 left only rows
df_cust_prod_ord = df_cust_prod_ord[df_cust_prod_ord['_merge']=='both']

In [37]:
#check to make sure count is 32435059
df_cust_prod_ord.shape

(32435059, 33)

In [38]:
df_cust_prod_ord.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329.0      1.0           1.0                 2.0                8.0   
1  2398795.0      1.0           2.0                 3.0                7.0   
2   473747.0      1.0           3.0                 3.0               12.0   
3  2254736.0      1.0           4.0                 4.0                7.0   
4   431534.0      1.0           5.0                 4.0               15.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                1.0        0.0   
1                    15.0         196                1.0        1.0   
2                    21.0         196                1.0        1.0   
3                    29.0         196                1.0        1.0   
4                    28.0         196                1.0        1.0   

  product_name  ...  first_name  last_name  Gender    STATE   Age date_joined  \
0         Soda  ...       Linda     Nguyen  Female  Alabama  31.0   2/17/2019   
1         Soda  ...       Linda     Nguyen  Female  Alabama  31.0   2/17/2019   
2         Soda  ...       Linda     Nguyen  Female  Alabama  31.0   2/17/2019   
3         Soda  ...       Linda     Nguyen  Female  Alabama  31.0   2/17/2019   
4         Soda  ...       Linda     Nguyen  Female  Alabama  31.0   2/17/2019   

  num_of_dependants  fam_status   income  _merge  
0               3.0     married  40423.0    both  
1               3.0     married  40423.0    both  
2               3.0     married  40423.0    both  
3               3.0     married  40423.0    both  
4               3.0     married  40423.0    both  

[5 rows x 33 columns]

### 5
Export dataframe

In [39]:
#export df_cust_prod_ord as pkl file
df_cust_prod_ord.to_pickle(os.path.join(path, 'prepared data', 'final_orders_products_customers.pkl'))